sdk/src/types/block/capabilities.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use alloc::boxed::Box;
+use core::marker::PhantomData;
+
+use derive_more::Deref;
+use packable::{error::UnpackErrorExt, prefix::BoxedSlicePrefix, Packable};
+
+/// A list of bitflags that represent capabilities.
+#[derive(Debug, Deref)]
+#[repr(transparent)]
+pub struct Capabilities<Flag> {
+    #[deref]
+    bytes: BoxedSlicePrefix<u8, u8>,
+    _flag: PhantomData<Flag>,
+}
+
+impl<Flag> Capabilities<Flag> {
+    pub(crate) fn from_bytes(bytes: BoxedSlicePrefix<u8, u8>) -> Self {
+        Self {
+            bytes,
+            _flag: PhantomData,
+        }
+    }
+
+    /// Returns a [`Capabilities`] with every possible flag disabled.
+    pub fn none() -> Self {
+        Self::default()
+    }
+
+    /// Returns whether every possible [`AddressCapabilityFlag`] is disabled.
+    pub fn is_none(&self) -> bool {
+        self.iter().all(|b| 0.eq(b))
+    }
+}
+
+impl<Flag: CapabilityFlag> Capabilities<Flag> {
+    /// Returns a [`Capabilities`] with every possible flag enabled.
+    pub fn all() -> Self {
+        let mut res = Self::default();
+        res.set_all();
+        res
+    }
+
+    /// Returns whether every possible [`AddressCapabilityFlag`] is enabled.
+    pub fn is_all(&self) -> bool {
+        Flag::all().all(|flag| self.has_capability(flag))
+    }
+
+    /// Enables every possible [`AddressCapabilityFlag`].
+    pub fn set_all(&mut self) -> &mut Self {
+        for flag in Flag::all() {
+            self.add_capability(flag);
+        }
+        self
+    }
+
+    /// Disabled every possible [`AddressCapabilityFlag`].
+    pub fn set_none(&mut self) -> &mut Self {
+        *self = Default::default();
+        self
+    }
+
+    /// Enables a given [`AddressCapabilityFlag`].
+    pub fn add_capability(&mut self, flag: Flag) -> &mut Self {
+        if self.bytes.len() <= flag.index() {
+            let mut v = Box::<[_]>::from(self.bytes.clone()).into_vec();
+            v.resize(flag.index() + 1, 0);
+            // Unwrap: safe because the indexes are within u8 bounds
+            self.bytes = v.into_boxed_slice().try_into().unwrap();
+        }
+        self.bytes[flag.index()] |= flag.as_byte();
+        self
+    }
+
+    /// Enables a given set of flags.
+    pub fn add_capabilities(&mut self, flags: impl IntoIterator<Item = Flag>) -> &mut Self {
+        for flag in flags {
+            self.add_capability(flag);
+        }
+        self
+    }
+
+    /// Enables a given set of flags.
+    pub fn with_capabilities(mut self, flags: impl IntoIterator<Item = Flag>) -> Self {
+        self.add_capabilities(flags);
+        self
+    }
+
+    /// Enables a given set of flags.
```

Comment:
Not a big deal but these 3 fns have the same doc, it kinda works but it's still a bit weird